backend/czi_hosted/data_common/cache.py

```diff
@@ -0,0 +1,219 @@
+            except Exception:
+                # necessary to hold the reader lock after an exception, since
+                # the release will occur when the context exits.
+                self.data_lock.w_demote()
+                raise
+
+        # demote the write lock to a read lock.
+        self.data_lock.w_demote()
+        return self.data
+
+    def release(self):
+        """Release the reader lock"""
+        self.data_lock.r_release()
+
+    def delete(self):
+        """Clear resources used by this cache item, if self.data has a cleanup method, call it, if not
+        delete the cached information"""
+        with self.data_lock.w_locked():
+            if self.data:
+                try:
+                    self.data.cleanup()
+                except AttributeError:
+                    pass
+                self.data = None
+
+    def attempt_delete(self):
+        """Delete, but only if the write lock can be immediately locked.  Return True if the delete happened"""
+        if self.data_lock.w_acquire_non_blocking():
+            if self.data:
+                try:
+                    self.data.cleanup()
+                    self.data = None
+                except AttributeError:
+                    self.data = None
+                except Exception:
+                    # catch all exceptions to ensure the lock is released
+                    pass
+
+                self.data_lock.w_release()
+                return True
+        else:
+            return False
+
+
+class CacheItemInfo(object):
+    """
+    This class stores a pointer to and metadata about the cached item. When the data is accessd the
+    update function is called to track the latest access
+    """
+    def __init__(self, cache_item: CacheItem(), timestamp):
+        self.cache_item = cache_item
+        self.last_access = timestamp
+        self.num_access = 1
+
+    def update_latest_cache_access(self):
+        self.last_access = time.time()
+        self.num_access +=1
+
+class CacheManager(object):
+    """An base class to manage cached data. This is intended to be used as a context manager
+    for handling api requests.  When the context is created, the data_adaptor is either loaded or
+    retrieved from a cache.  In either case, the reader lock is taken during this time, and release
+    when the context ends.  This class currently implements a simple least recently used cache,
+    which can delete data from the cache to make room for a new one.
+
+    This is the intended usage pattern:
+
+           cache_manager = DataCacheManager(max_cached=..., timelimmit_s = ...)
+           with cache_manager.data_adaptor(location, app_config) as data_adaptor:
+               # use the data_adaptor for some operation
+    """
+
+    # FIXME:   If the number of active datasets exceeds the max_cached, then each request could
+    # lead to a dataset being deleted and a new only being opened: the cache will get thrashed.
```

Comment:
I think the number of active datasets(in prod) is less than the max_cached, also I suspect our access may be somewhat bursty with lots of people accessing the same dataset at once (for example after a publication or author tweet). In the long run considering dataset size in addition to number of datasets for the caching of the actual dataset may be worth considering.  Because this is actively blocking the revisions workI think a redesign using the python tools is a little out of scope but would be good to keep in mind for the future